src/utils/recompiler.ts

```diff
@@ -0,0 +1,134 @@
+import fs from "fs";
+import path from "path";
+import { createHash } from "crypto";
+import { HardhatRuntimeEnvironment, TaskArguments } from "hardhat/types";
+
+const fsPromises = fs.promises;
+const ok = Object.keys;
+
+// FileName - HashPair
+export interface NameHashPair {
+    [key: string]: string;
+}
+
+// Save FileNameHashPair
+const nameHashPair: NameHashPair = {};
+
+// Default Source Path and Target Paths
+let defaultSourcesPath: string;
+// let sourcesPaths: string[];
+
+// Set to save all changed contracts or
+// artifacts that are not available
+const changed = new Set();
+
+// File containing contract hash
+let tracker: NameHashPair = {};
+
+// Creates tracker file only if it doesn't exist in cache
+const upsertFile = async () => {
+    const dirpath = path.join(defaultSourcesPath, "../cache/cairo-files-cache.json");
+    const cacheDirpath = path.join(defaultSourcesPath, "../cache");
+
+    // Creates cache directory if it doesn't exist
+    if (!fs.existsSync(cacheDirpath)) {
+        fs.mkdirSync(cacheDirpath);
+    }
+
+    if (!fs.existsSync(dirpath)) {
+        // create file, if it's not found
+        await fsPromises.writeFile(dirpath, JSON.stringify({}));
+    } else {
+        // try to read file
+        const oldNameHashPair = await fsPromises.readFile(dirpath, "utf8");
+        tracker = JSON.parse(oldNameHashPair || "{}");
+    }
+};
+
+// Gets hash of each .cairo file inside contracts
+const getContractHash = async () => {
+    try {
+        // traverse directory contracts/
+        const dirpath = path.join(defaultSourcesPath, "../contracts");
+        const files = await fsPromises.readdir(dirpath);
+        // check only cairo file extensions
+        const filesList = files.filter((el) => path.extname(el).toLowerCase() === ".cairo");
+        // select file name
+        for (const cairoContract of filesList) {
+            const data = await fsPromises.readFile(dirpath.concat(`/${cairoContract}`));
+            const hash = createHash("sha256");
+            hash.update(data);
+            nameHashPair[`${cairoContract}`] = hash.copy().digest("hex").toString();
+        }
+    } catch (err) {
+        console.log(err);
+    }
+};
+
+// Checks artifacts availability
+const checkArtifacts = async () => {
+    try {
+        // traverse directory starknet-artifacts/contracts
+        const dirpath = path.join(defaultSourcesPath, "../starknet-artifacts/contracts");
+        const files = await fsPromises.readdir(dirpath);
+        for (const name of ok(nameHashPair)) {
+            if (!files.includes(name)) {
+                changed.add(`${defaultSourcesPath}/${name}`);
+            }
+        }
+    } catch (err) {
+        console.log(err);
+    }
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (hre: HardhatRuntimeEnvironment): Promise<boolean> => {
+    if (ok(nameHashPair).length === ok(tracker).length) {
```

Comment:
Just relying on comparing the lengths is not a good idea. Apart from it not being clear where `nameHashpair` is populated (since it is a global var), if I understand correctly, this `if` would be sastisfied if e.g. one file was deleted and a new one was added.

EDIT: Actually looking at it, don't `if` and `else` branches do the same thing here?

Reply:
Yes, they do the same thing. Need to fix that as well.

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/106/commits/585b8eff4df386f762ec754b7fa56f631bc223ec